src/sapling/sapling_validation.cpp

```diff
@@ -16,6 +16,113 @@
 
 namespace SaplingValidation {
 
+// Verifies that Shielded txs are properly formed and performs content-independent checks
+bool CheckTransaction(const CTransaction& tx, CValidationState& state, CAmount& nValueOut, bool fIsSaplingActive)
+{
+    bool hasSaplingData = tx.hasSaplingData();
+
+    // v1 must not have shielded data.
+    if (tx.nVersion < CTransaction::SAPLING_VERSION && hasSaplingData) {
+        return state.DoS(100, error("%s: Not Sapling version with Sapling data", __func__ ),
+                         REJECT_INVALID, "bad-txns-form-not-sapling");
+    }
+
+    // From here, all of the checks are done in +v2 transactions.
```

Comment:
Actually we can still have a v1 transaction (without sapData) here. Only after line 43 we can be sure we are  only dealing with v2+  txes.
Styling (mostly irrelevant) nit/suggestion: in fact, i think it would be clearer if you switch the order of the checks... you could even get rid of one `if (hasSaplingData)` block:
from
```cpp
    // From here, all of the checks are done in +v2 transactions.

    // if the tx has shielded data, cannot be a coinstake, coinbase, zcspend, p2csOut and zcmint
    if (hasSaplingData) {
        if (tx.IsCoinStake() || ... ) return state.DoS();
    }

    // +v2, if the tx has no shielded data, return true. No check needed.
    if (!hasSaplingData)  return true;
```

to:
```cpp
     // No shielded data. Nothing to check.
    if (!hasSaplingData) return true;

    // From here on, all of the checks are done on +v2 transactions with saplingData.

    // if the tx has shielded data, cannot be a coinstake, or...
    if (tx.IsCoinStake() || ... ) return state.DoS(...);
```